replica/table.cc

```diff
@@ -193,11 +211,7 @@ table::make_reader_v2(schema_ptr s,
     // https://github.com/scylladb/scylla/issues/309
     // https://github.com/scylladb/scylla/issues/185
 
-    for (auto&& mt : *_compaction_group->memtables()) {
-        if (auto reader_opt = mt->make_flat_reader_opt(s, permit, range, slice, pc, trace_state, fwd, fwd_mr)) {
-            readers.emplace_back(std::move(*reader_opt));
-        }
-    }
+    add_memtables_to_reader_list(readers, s, permit, range, slice, pc, trace_state, fwd, fwd_mr);
```

Comment:
Here, we should either select one reader (single partition), or create a virtual reader that stitches together readers for relevant memtables (scans). The scan case should grab the memtable at the last possible moment, to avoid pinning it in memory.  

In fact the loop around compaction groups should enclose the cache readers (and create a new cache reader every time we cross a compaction group boundary).

Reply:
Yes, I forgot the details here. A reader_selector should be enough.

btw, we'll have to get rid of it once we have tablets (also the multishard reader) since a tablet will map to just one shard. That means the role of the multishard reader will be assumed by storage_proxy::query_partition_key_range_concurrent.

Reply:
@avikivity why can't we reuse the same cache reader? why do we have to create a new one when moving to next group? once we cross compaction group boundary, the cache reader can still feed the data for the current position.

similar to what happens when L0 sstables are read along with Ln sstables, where n > 0. L0 sstables span roughly the entire token range, meaning their readers will be alive most of the time.

combined reader
   incremental memtable reader selection (assumes disjointness)
   incremental sstable reader selection (assumes disjointness across groups)
   row cache reader

Reply:
We can use the same reader, with an incremental selector.

The downside is that it pins sstables from all compaction groups, but it's not a regression from what we have now (where we pin all sstables anyway). No regression -> no requirement.